tests/unit/test_grub.py

```diff
@@ -358,144 +394,233 @@ def test_remove_no_config(self, mock_save, mock_apply):
-            self.name: EXP_GRUB_CONFIG,
-            "charm-b": {"GRUB_TIMEOUT": "0"},
-            "charm-c": {
-                "GRUB_TERMINAL": "console",
-                "GRUB_CMDLINE_LINUX_DEFAULT": '"$GRUB_CMDLINE_LINUX_DEFAULT hugepagesz=1G"',
-            },
-        }
-        self.load_config.side_effect = [exp_configs["charm-b"], exp_configs["charm-c"]]
-        (self.tmp_dir / "90-juju-charm-b").touch()
-        (self.tmp_dir / "90-juju-charm-c").touch()
+    def test_remove(self, mock_save, mock_applied_configs, mock_apply):
+        """Test removing with applying."""
+        self.config.path.touch()  # created charm file
+        mock_applied_configs.values.return_value = [
+            {key: value for key, value in EXP_GRUB_CONFIG.items() if key != "GRUB_TIMEOUT"}
+        ]
 
         changed_keys = self.config.remove()
 
-        self.assertFalse((self.tmp_dir / self.name).exists())
-        self.load_config.assert_has_calls(
-            [
-                mock.call(self.tmp_dir / f"90-juju-{charm}")
-                for charm in exp_configs
-                if charm != self.name
-            ]
-        )
-        self.assertSetEqual(changed_keys, exp_changed_keys)
-        self.assertDictEqual(
-            self.config._data,
-            {
-                "GRUB_TIMEOUT": "0",
-                "GRUB_TERMINAL": "console",
-                "GRUB_CMDLINE_LINUX_DEFAULT": '"$GRUB_CMDLINE_LINUX_DEFAULT hugepagesz=1G"',
-            },
-        )
-        mock_save.assert_called_once()
-        mock_apply.assert_called_once()
+        self.assertFalse(self.config.path.exists())
+        self.assertSetEqual(changed_keys, {"GRUB_TIMEOUT"})
+        mock_save.assert_called_once_with()
+        mock_apply.assert_called_once_with()
 
     @mock.patch.object(grub.Config, "apply")
     @mock.patch.object(grub.Config, "_save_grub_configuration")
-    def test_update_on_container(self, mock_save, mock_apply):
+    @mock.patch.object(grub, "_update_config")
+    def test_update_on_container(self, mock_update, mock_save, mock_apply):
         """Test update current GRUB config on container."""
         self.is_container.return_value = True
 
         with self.assertRaises(grub.IsContainerError):
             self.config.update({"GRUB_TIMEOUT": "0"})
 
+        self.assertDictEqual(self.config._data, EXP_GRUB_CONFIG, "config was changed")
         mock_save.assert_not_called()
         mock_apply.assert_not_called()
-        self.assertDictEqual(self.config._data, EXP_GRUB_CONFIG, "config was changed")
+        mock_update.assert_not_called()
 
     @mock.patch.object(grub.Config, "apply")
     @mock.patch.object(grub.Config, "_save_grub_configuration")
-    @mock.patch.object(
-        grub.Config, "_set_value", side_effect=grub.ValidationError("test", "test_message")
-    )
-    def test_update_validation_failure(self, _, mock_save, mock_apply):
+    @mock.patch.object(grub.Config, "_update")
+    @mock.patch.object(grub, "_update_config")
```

Comment:
I still don't like all this mocking. It makes the tests more fragile as we're testing a lot more test internals. Ideally we could change these to use the real filesystem (in a temp directory) where possible, and only use mocking where necessary (for example, the `subprocess` update-grub calls). Local filesystem operations are fast and reliable, so don't need to be mocked.

Reply:
Hi @benhoyt, I agree, but at the same time the definition of unit tests is testing functionality of single function, because it's not good if I change one thing and it breaks half of unit tests. So for me it's really necessary to tests each function deeply with mocked everything else.
However at the beginning I mention that I agree with you and that's why I provide `TestSmokeConfig` where I mocked everything and `TestFullConfig` where I mocked only subprocess check_call. I think that it is good solution, because like this I test all scenarios for each function with mocking other functions and the general usage for `.remove` and `.update` function.
What do you think?

Reply:
> I agree, but at the same time the definition of unit tests is testing functionality of single function, because it's not good if I change one thing and it breaks half of unit tests. So for me it's really necessary to tests each function deeply with mocked everything else.

"Unit testing" is a somewhat vague term, but just given the name, I don't think it necessarily means testing functions, but testing *units* (which could be modules, classes, methods). That said, I don't believe that debating the term is particularly helpful. I agree with you that we want to avoid "if I change one thing it breaks half of the tests".

But, as I see it, mocking internal helper functions is exactly how you'd get that problem. This test is mocking the internal `_save_grub_configuration`, `_update`, and `_update_config` functions. But those are implementation details, so if we change the name or implementation of any of those methods, this test will break. And that's fragile. That's why I advocate mocking as little as possible, and -- as much as possible -- only testing the external contract, not the implementation details. On the other hand, we probably do need to mock `subprocess.check_call` because it's an external service/program.

In any case, those are my thoughts. I'm not going to block this PR because the testing is not too my taste. We'll see if the tests end up being fragile and we can always change the testing approach later. And we can keep discussing what "unit testing" means over a beer in Riga. :-)